### ISSUE TITLE
[apex] New Rule: Avoid using `instanceof Id`

### ISSUE THREAD

**Issue by mitchspano** [pmd/pmd#6211]
**Proposed Rule Name:** `AvoidInstanceOfId`

**Proposed Category:** Error Prone

**Description:**

The `instanceof Id` operator in Apex performs a very loose and unreliable check. It primarily seems to verify string length and potentially specific character positions, rather than validating the entire character set required for a Salesforce Id.

This can lead to `instanceof Id` incorrectly evaluating to true for strings that contain invalid characters. This behavior is inconsistent, as attempting to assign or cast that same invalid string to an Id variable will throw a `System.StringException`. Relying on `instanceof Id` can lead to unexpected runtime errors

**Code Sample:** 

```java
// instanceOf is very unreliable
System.debug('xxxx0xxxxxxxxxx' instanceof Id); // true 
System.debug('xxxx0xxxxxxxx?!' instanceof Id); // true
System.debug('?!@_0-&<>/{}][)' instanceof Id); // true (15 chars)
System.debug('xxxx0xxxxxxxx?' instanceof Id); // false (16 chars)
System.debug('xxxx1xxxxxxxxxx' instanceof Id); // false (char at index 4 != '0')

String bogusId = 'c&*()b02@!#$%^.com'; // clearly a bogus Id
System.debug(bogusId instanceof Id); // true
Id foo = bogusId; // throws StringException

String oddId = 'xxxx1xxxxxxxxxx'; // satisfies published format specifications for an Id
System.debug(oddId instanceof Id); // true
Id bar = oddId; // throws StringException
```

There is no built-in, reliable static method to simply check if a given string could be a valid Id according to the character set and length rules without throwing an exception. The only relatively safe and reliable way currently is to use a try-catch block around a cast/assign.

```java
public static Boolean isValidId(String s) { 
  try {
    Id unused = (Id) s; // or `Id.valueOf(s) in API 54.0 or higher`
  } catch (StringException e) {
    return false;
  }
  return true;
}
```

Any call of `instanceof Id` should be flagged as problematic.

**Possible Properties:**

*   Should this rule be customizable via properties?
No need for properties
